deepinv/datasets/base.py

```diff
@@ -0,0 +1,383 @@
+
+    Datasets should ideally return :class:`torch.Tensor` or :class:`deepinv.utils.TensorList` so that they are batchable and can be used with `deepinv`.
+
+    If using DeepInverse with your own custom dataset, you should inherit from this class and use :func:`check_dataset` to check your dataset is compatible.
+    """
+
+    def check_dataset(self):
+        """Check dataset returns correct format of images or image tuples."""
+        check_dataset(self, allow_non_tensor=True)
+
+    def __len__(self):
+        raise NotImplementedError()
+
+    def __getitem__(self, idx: int):
+        raise NotImplementedError()
+
+
+class TensorDataset(ImageDataset):
+    r"""
+    Dataset wrapping data explicitly passed as tensors.
+
+    This dataset can be used to return ground truth `x`, ground truth and measurements `(x, y)`, or measurements only `(y)`.
+    All input tensors must be of shape `(N, ...)` and of same `N` where N is the number of samples and ... represents the data dimensions.
+
+    Optionally, `params` are returned too.
+
+    :param torch.Tensor, None x: optional input ground truth tensor `x`
+    :param torch.Tensor, None y: optional input measurement tensor `y`
+    :param dict[str, torch.Tensor], None params: optional input physics parameters `params` of format `{"str": Tensor}`
+
+    |sep|
+
+    Examples:
+
+    Construct a dataset from a single measurement only:
+
+    >>> import torch
+    >>> from deepinv.datasets import TensorDataset
+    >>> y = torch.rand(1, 3, 8, 8) # B,C,H,W
+    >>> dataset = TensorDataset(y=y)
+    >>> x, y = dataset[0]
+    >>> x
+    nan
+    >>> y.shape
+    torch.Size([3, 8, 8])
+
+    Construct a dataset from a ground truth batch:
+
+    >>> x = torch.rand(4, 3, 8, 8)  # 4 samples of 3-channel 8x8 images
+    >>> dataset = TensorDataset(x=x)
+    >>> dataset[0].shape
+    torch.Size([3, 8, 8])
+
+
+    """
+
+    def __init__(
+        self,
+        *,
+        x: Optional[Tensor] = None,
+        y: Optional[Tensor] = None,
+        params: Optional[dict[str, Tensor]] = None,
+    ):
+        self.x = x
+        self.y = y
+        self.params = params
+
+        if (
+            isinstance(x, CORE_TYPES)
+            and isinstance(y, CORE_TYPES)
+            and x.size(0) != y.size(0)
+        ):
+            raise ValueError(
+                f"x must be same size as y in dimension 0, but got {x.size(0)} vs {y.size(0)}"
+            )
+        elif self.check_none(x) and self.check_none(y):
+            raise ValueError("At least one of x or y must be not None or not nan.")
+        elif not self.check_none(x) and not isinstance(x, CORE_TYPES):
+            raise ValueError("x must be Tensor or TensorList.")
+        elif not self.check_none(y) and not isinstance(y, CORE_TYPES):
+            raise ValueError("y must be Tensor or TensorList.")
+
+    def check_none(self, x) -> bool:
```

Comment:
```suggestion
    def _is_none_or_nan(self, x) -> bool:
```

This function does not check if a tensor/tensorlist full of nan?

Reply:
No, a tensor/tensorlist full of nan is not nan, it is a tensor/tensorlist, which I guess is a valid signal for `x`. `x is nan` only if `math.isnan(x)` which captures both a math nan and a single-element tensor nan `tensor(nan)` or `tensor([nan])`. In these cases, `x` represents a lack of ground truth.

Reply:
Any further discussion can be continued in #674